server/src/main/java/org/apache/druid/server/http/DataSourcesResource.java

```diff
@@ -235,8 +231,8 @@ public Response deleteDataSource(
                        .build();
       }
     } else {
-      if (!databaseSegmentManager.removeDataSource(dataSourceName)) {
-        return Response.noContent().build();
+      if (!segmentsMetadata.tryMarkAsUnusedAllSegmentsInDataSource(dataSourceName)) {
+        return Response.serverError().build();
```

Comment:
Http status code "NO_CONTENT(204)"  might be appropriate in some cases instead of "INTERNAL_SERVER_ERROR(500)"  as some methods in `SQLMetadataSegmentManager.java` return false if nothing gets updated in the database, which is not necessarily a server error. I was looking to return appropriate http codes when working on [this issue](https://github.com/apache/incubator-druid/issues/7439)

Reply:
> which is not necessarily a server error

It looks to me like a server error, can you elaborate when it isn't?

Reply:
so, for example, `SQLMetadataSegmentManager #removeSegmentFromTable` returns `removed > 0` which is in turn called by `removeSegment(SegmentId segmentId)`, my doubt is what if nothing gets updated and it returns 0 and there is no other exception. Should the api return server error in such cases ?

Reply:
@surekhasaharan right. I've changed the bahaviour of these REST endpoints to return normally
`{"segmentStateChanged": true/false}`
and
`{"numChangedSegments": <number>}`
on idempotent operations.

I think this would also allow detecting non-existent segments or data source names because they return `{"segmentStateChanged": false}` and `{"numChangedSegments": 0}` respectively.

On the other hand, database errors are still exposed as 500 internal server errors.